src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -40149,6 +40151,579 @@ public void CopyTupleUnconstrainedElementNullability_02()
+        public void Conversions_ImplicitNullable_03()
+        {
+            var source =
+@"class Program
+{
+    static void F(int x)
+    {
+        int? y = x;
+        long? z = x;
+        y.Value.ToString();
+        z.Value.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(32531, "https://github.com/dotnet/roslyn/issues/32531")]
+        public void Conversions_ImplicitNullable_04()
+        {
+            var source =
+@"class Program
+{
+    static void F1(long x1)
+    {
+        int? y1 = x1; // 1
+        y1.Value.ToString();
+    }
+    static void F2(long? x2)
+    {
+        int? y2 = x2; // 2
+        y2.Value.ToString(); // 3
+    }
+    static void F3(long? x3)
+    {
+        if (x3 == null) return;
+        int? y3 = x3; // 4
+        y3.Value.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (5,19): error CS0266: Cannot implicitly convert type 'long' to 'int?'. An explicit conversion exists (are you missing a cast?)
+                //         int? y1 = x1; // 1
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "x1").WithArguments("long", "int?").WithLocation(5, 19),
+                // (10,19): error CS0266: Cannot implicitly convert type 'long?' to 'int?'. An explicit conversion exists (are you missing a cast?)
+                //         int? y2 = x2; // 2
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "x2").WithArguments("long?", "int?").WithLocation(10, 19),
+                // (11,9): warning CS8629: Nullable value type may be null.
+                //         y2.Value.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "y2.Value").WithLocation(11, 9),
+                // (16,19): error CS0266: Cannot implicitly convert type 'long?' to 'int?'. An explicit conversion exists (are you missing a cast?)
+                //         int? y3 = x3; // 4
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "x3").WithArguments("long?", "int?").WithLocation(16, 19));
+        }
+
+        [Fact]
+        [WorkItem(32531, "https://github.com/dotnet/roslyn/issues/32531")]
+        public void Conversions_ImplicitNullable_05()
+        {
+            var source =
+@"#pragma warning disable 0649
+struct A
+{
+    internal object? F;
+}
+struct B
+{
+    internal object? F;
+}
+class Program
+{
+    static void F1()
+    {
+        A a1 = new A();
+        B? b1 = a1; // 1
+        _ = b1.Value;
+        b1.Value.F.ToString(); // 2
+    }
+    static void F2()
+    {
+        A? a2 = new A() { F = 2 };
+        B? b2 = a2; // 3
+        _ = b2.Value;
+        b2.Value.F.ToString(); // 4
+    }
+    static void F3(A? a3)
+    {
+        B? b3 = a3; // 5
+        _ = b3.Value; // 6
```

Comment:
>_ = b3.Value; // 6 [](start = 8, length = 18)

It feels inconsistent that in this case we report a warning, but in two cases above we don't. It doesn't look like there is much difference in all these scenarios, all of them about lack of conversion between types.

Reply:
`a3` may be `null` where as `a1` and `a2` are not.

---
In reply to: [251101402](https://github.com/dotnet/roslyn/pull/32729#discussion_r251101402) [](ancestors = 251101402)

Reply:
> a3 may be null where as a1 and a2 are not.

And we know that why?

---
In reply to: [251103673](https://github.com/dotnet/roslyn/pull/32729#discussion_r251103673) [](ancestors = 251103673,251101402)

Reply:
`a1` and `a2` are initialized with `new A()`.

---
In reply to: [251104474](https://github.com/dotnet/roslyn/pull/32729#discussion_r251104474) [](ancestors = 251104474,251103673,251101402)